projects/slap_manipulation/src/slap_manipulation/optim/lamb.py

```diff
@@ -0,0 +1,130 @@
+"""Lamb optimizer."""
+
+# LAMB optimizer used as is.
+# 2ndary source: https://github.com/peract/peract/blob/main/helpers/optim/lamb.py
+# Source: https://github.com/cybertronai/pytorch-lamb
+# License: https://github.com/cybertronai/pytorch-lamb/blob/master/LICENSE
+# Note: Also available at https://github.com/jettify/pytorch-optimizer
+
+import collections
+import math
+
+import torch
+from torch.optim import Optimizer
+
+# def log_lamb_rs(optimizer: Optimizer, event_writer: SummaryWriter, token_count: int):
+#     """Log a histogram of trust ratio scalars in across layers."""
+#     results = collections.defaultdict(list)
+#     for group in optimizer.param_groups:
+#         for p in group['params']:
+#             state = optimizer.state[p]
+#             for i in ('weight_norm', 'adam_norm', 'trust_ratio'):
+#                 if i in state:
+#                     results[i].append(state[i])
+#
+#     for k, v in results.items():
+#         event_writer.add_histogram(f'lamb/{k}', torch.tensor(v), token_count)
+
+
+class Lamb(Optimizer):
```

Comment:
where is this code from? check license?

Reply:
Directly taken from peract which takes it as is from `cybertronai/pytorch-lamb`. License MIT: https://github.com/cybertronai/pytorch-lamb/blob/master/LICENSE

Reply:
add a link in comment and include the license if you copy the code from it